legion/engine/llri-vk/detail/adapter.cpp

```diff
@@ -74,6 +74,78 @@ namespace llri
         return false;
     }
 
+    result Adapter::impl_querySurfacePresentSupportEXT(SurfaceEXT* surface, queue_type type, bool* support) const
+    {
+        auto queueFamilies = internal::findQueueFamilies(static_cast<VkPhysicalDevice>(m_ptr));
+
+        if(queueFamilies[type] == UINT_MAX)
+        {
+            *support = false;
+            return result::Success;
+        }
+        
+        VkBool32 supported;
+        const auto r = vkGetPhysicalDeviceSurfaceSupportKHR(static_cast<VkPhysicalDevice>(m_ptr), queueFamilies[type], static_cast<VkSurfaceKHR>(surface->m_ptr), &supported);
+        
+        if (r != VK_SUCCESS)
+            return internal::mapVkResult(r);
+
+        *support = static_cast<bool>(supported);
+        return result::Success;
+    }
+
+    result Adapter::impl_querySurfaceCapabilitiesEXT(SurfaceEXT* surface, surface_capabilities_ext* capabilities) const
+    {
+        VkResult r;
+        uint32_t count;
+
+        // handle formats
+        r = vkGetPhysicalDeviceSurfaceFormatsKHR(static_cast<VkPhysicalDevice>(m_ptr), static_cast<VkSurfaceKHR>(surface->m_ptr), &count, nullptr);
+        if (r != VK_SUCCESS)
+            return internal::mapVkResult(r);
+        std::vector<VkSurfaceFormatKHR> formats(count);
+        r = vkGetPhysicalDeviceSurfaceFormatsKHR(static_cast<VkPhysicalDevice>(m_ptr), static_cast<VkSurfaceKHR>(surface->m_ptr), &count, formats.data());
+        if (r != VK_SUCCESS)
+            return internal::mapVkResult(r);
+
+        capabilities->formats = std::vector<format>(count);
```

Comment:
shouldn't formats have already been initialized? so technically you should just be able to do a resize? avoiding a copy. also possibly avoiding a realloc or memzero or calling destructors and constructors for each element, depending on what kind of type format is. especially since you're overwriting all the values anyways

Reply:
fair, will fix